In [18]:
import pandas as pd
import pickle

from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler, PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, ElasticNet
from sklearn.pipeline import make_pipeline

__Import des données dans le dataframe "df" et paramétrage des options__

In [19]:
df = pd.read_csv("kc_house_data.modelisation.csv")

__Création du dataframe et de la series, puis des données de test et de train__

In [20]:
X = df.drop(["id", "date", "prix"], axis=1)
X2 = df.drop(["id", "date", "prix", "m2_habitable15", "m2_parcelle15"], axis=1)
y = df["prix"]

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y, random_state=42)

__Test - Régression linéaire__

In [21]:
regressor = LinearRegression()

regressor.fit(X_train, y_train)
regressor.score(X_test, y_test)

0.7969470445258322

__Test - Robust + Polynomial + Lasso__

In [22]:
model_robust_lasso = make_pipeline(RobustScaler(), PolynomialFeatures(), Lasso(alpha=0.1))

model_robust_lasso.fit(X_train, y_train)
model_robust_lasso.score(X_test, y_test)

/home/alex/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 7.686e+13, tolerance: 2.171e+11
  model = cd_fast.enet_coordinate_descent(


0.9098598044862317

__Test - Robust + Polynomial + ElasticNet__

In [23]:
model_robust_elastic = make_pipeline(RobustScaler(), PolynomialFeatures(), ElasticNet())

model_robust_elastic.fit(X_train, y_train)
model_robust_elastic.score(X_test, y_test)

/home/alex/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.748e+14, tolerance: 2.171e+11
  model = cd_fast.enet_coordinate_descent(


0.7884506788914828

__Test - MinMax + Polynomial + Lasso__

In [24]:
model_minmax_lasso = make_pipeline(MinMaxScaler(), PolynomialFeatures(), Lasso(alpha=0.1))

model_minmax_lasso.fit(X_train, y_train)
model_minmax_lasso.score(X_test, y_test)

/home/alex/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.016e+13, tolerance: 2.171e+11
  model = cd_fast.enet_coordinate_descent(


0.9100710826478696

__Test - MinMax + Polynomial + Elastic__

In [25]:
model_minmax_elastic = make_pipeline(MinMaxScaler(), PolynomialFeatures(), ElasticNet())

model_minmax_elastic.fit(X_train, y_train)
model_minmax_elastic.score(X_test, y_test)

0.39398679032975314

__Test - Standard + Polynomial + Lasso__

In [26]:
model_standard_lasso = make_pipeline(StandardScaler(), PolynomialFeatures(), Lasso(alpha=0.1))

model_standard_lasso.fit(X_train, y_train)
model_standard_lasso.score(X_test, y_test)

/home/alex/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 7.680e+13, tolerance: 2.171e+11
  model = cd_fast.enet_coordinate_descent(


0.7961781157724125

__Test - Standard + Polynomial + Elastic__

In [27]:
model_standard_elastic = make_pipeline(StandardScaler(), PolynomialFeatures(), ElasticNet())

model_standard_elastic.fit(X_train, y_train)
model_standard_elastic.score(X_test, y_test)

0.8943381994321498

__Test - Robust + Polynomial + Lasso - Sans les 15 plus proches voisins__

In [28]:
model_robust_lasso_2 = make_pipeline(RobustScaler(), PolynomialFeatures(), Lasso(alpha=0.1))

model_robust_lasso_2.fit(X2_train, y2_train)
model_robust_lasso_2.score(X2_test, y2_test)

/home/alex/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.114e+13, tolerance: 2.171e+11
  model = cd_fast.enet_coordinate_descent(


0.9079787829451107

__Test - MinMax + Polynomial + Lasso - Sans les 15 plus proches voisins__

In [29]:
model_minmax_lasso_2 = make_pipeline(MinMaxScaler(), PolynomialFeatures(), Lasso(alpha=0.1))

model_minmax_lasso_2.fit(X2_train, y2_train)
model_minmax_lasso_2.score(X2_test, y2_test)

/home/alex/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.448e+13, tolerance: 2.171e+11
  model = cd_fast.enet_coordinate_descent(


0.9067740135463744

__Meilleurs résultats :__\
MinMax + Polynomail + Lasso : 0.9100710826478696\
Robust + Polynomial + Lasso : 0.9098598044862317

__Sans les 15 plus proches voisins :__\
Robust + Polynomial + Lasso : 0.9079787829451107\
MinMax + Polynomial + Lasso : 0.9067740135463744

__Export du model via pickle__

In [30]:
"""pickle_out = open("Model.pkl", "wb")
pickle.dump(model_robust_lasso_2, pickle_out)
pickle_out.close()"""